src/mlpack/methods/reinforcement_learning/q_learning.hpp

```diff
@@ -0,0 +1,170 @@
+ *               Alex Graves and
+ *               Ioannis Antonoglou and
+ *               Daan Wierstra and
+ *               Martin A. Riedmiller},
+ *  title     = {Playing Atari with Deep Reinforcement Learning},
+ *  journal   = {CoRR},
+ *  year      = {2013},
+ *  url       = {http://arxiv.org/abs/1312.5602}
+ * }
+ * @endcode
+ *
+ * @tparam EnvironmentType The environment of the reinforcement learning task.
+ * @tparam NetworkType The network to compute action value.
+ * @tparam OptimizerType The optimizer to train the network.
+ * @tparam PolicyType Behavior policy of the agent.
+ * @tparam ReplayType Experience replay method.
+ */
+template <
+  typename EnvironmentType,
+  typename NetworkType,
+  typename OptimizerType,
+  typename PolicyType,
+  typename ReplayType = RandomReplay<EnvironmentType>
+>
+class QLearning
+{
+ public:
+  //! Convenient typedef for state.
+  using StateType = typename EnvironmentType::State;
+
+  //! Convenient typedef for action.
+  using ActionType = typename EnvironmentType::Action;
+
+  /**
+   * Create the QLearning object with given settings.
+   *
+   * @param network The network to compute action value.
+   * @param optimizer The optimizer to train the network.
+   * @param discount Discount for future return.
+   * @param policy Behavior policy of the agent.
+   * @param replayMethod Experience replay method.
+   * @param targetNetworkSyncInterval Interval (steps) to sync the target network.
+   * @param explorationSteps Steps before starting to learn.
+   * @param doubleQLearning Whether to use double Q-Learning.
+   * @param stepLimit Maximum steps in each episode, 0 means no limit.
+   * @param environment Reinforcement learning task.
+   */
+  QLearning(NetworkType& network,
+            OptimizerType& optimizer,
+            double discount,
+            PolicyType policy,
+            ReplayType replayMethod,
+            size_t targetNetworkSyncInterval,
+            size_t explorationSteps,
+            bool doubleQLearning = false,
+            size_t stepLimit = 0,
+            EnvironmentType environment = EnvironmentType());
+
+  /**
+   * Execute a step in an episode.
+   * @return Reward for the step.
+   */
+  double Step();
+
+  /**
+   * Execute an episode.
+   * @return Return of the episode.
+   */
+  double Episode();
+
+  /**
+   * @return Total steps from beginning.
+   */
+  const size_t& TotalSteps() const { return totalSteps; }
+
+  //! Modify the training mode / test mode indicator.
+  bool& Deterministic() { return deterministic; }
+
+  //! Get the indicator of training mode / test mode.
+  const bool& Deterministic() const { return deterministic; }
+
+ private:
+  /**
+   * Select the best action based on given action value.
+   * @param actionValues Action values.
+   * @return Selected actions.
+   */
+  arma::icolvec BestAction(const arma::mat& actionValues);
+
+  //! Reference of the learning network.
+  NetworkType& learningNetwork;
```

Comment:
I suspect that any development here will be part of 3.0.0, not 2.2.x---I think that only backported fixes will go into 2.2.x at this point really.

I think that the second solution is better here; in my view I don't see a reason to bind the `FunctionType` to the `OptimizerType` instance.  Instead I agree with Shangtong and think the design would be better as:

```
class GradientDescent
{
 public:
  GradientDescent(...);

  template<typename FunctionType>
  void Optimize(FunctionType& function, arma::mat& iterate);

  ...
};
```

I think that this refactoring would also make @micyril's hyperparameter tuner easier to design; a little more background on that in the second half of this email:

http://knife.lugatgt.org/pipermail/mlpack/2017-May/003327.html

When I originally made the `OptimizerType` idea, I thought that it was necessary to bind the function so that the optimizer could use `GetInitialPoint()` for the start of the optimization.  But as time has gone on, I think that `GetInitialPoint()` is not used internally in any optimizer, so now there is no reason to bind the optimizer type to the function type that it is optimizing.

I guess, to summarize, I think option 2 is the right way to do this.  If we all agree I can open a ticket for it, and help with the refactoring if needed.  (It should be a fairly simple refactoring, might take a little bit of time though.)